hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RegionScannerImpl.java

```diff
@@ -608,6 +610,11 @@ private boolean nextInternal(List<Cell> results, ScannerContext scannerContext)
           return scannerContext.setScannerState(NextState.NO_MORE_VALUES).hasMoreValues();
         }
         if (!shouldStop) {
+          // Read nothing as the cells were filtered, but still need to check time limit.
```

Comment:
Why we need to add a check here?

Reply:
To be honest, I think this should have been here all along and was just missed along the way. I'm not sure why we'd want to check time limit for the nextRow calls above but not this one. This check here ensures that populating from joined heap + nextRow does not exceed time or (new) size limit.

Reply:
This method is on the critial path of reading, and the code here will executed every time when we get a row, so it may affect scan performance if we add more checks here.
I just mean is it a must to have a check here? Why we do not need to check here in the past...

Reply:
Actually, this section of the method is not nearly as hot as the rest of the method. The only real way we reach this point is when `filter.filterRowCells(kvs)` clears all cells from the results after having been accumulated in StoreScanner.  There are only 2 standard filters which do this -- DependentColumnFilter and SingleColumnValueExcludeFilter.

That said, you do make a good point.  We have never had a time limit here, so we may not need it. We _do_ need a size limit check here, now that we track block sizes. Previously, we would not check size limit here because the results are empty so wouldn't have accumulated size progress. Now that we accumulate block size progress even for filtered rows, we need a check. 

For this type of scan, we will have accumulated blocks in both `populateResults()` and possibly `nextRow()`. Right after `populateResults()` there's a `scannerContext.checkAnyLimitReached(LimitScope.BETWEEN_CELLS)` call. That call doesn't protect against this case, because it passes `BETWEEN_CELLS`.  For scans with `filter.hasFilterRow()`, the limit scope is changed to `LimitScope.BETWEEN_ROWS`. So this check is skipped for these.  Scans which enter this code block will have skipped all other limit checks above. The checkSizeLimit I add here is the only safe place we can check BETWEEN_ROWS for these types of filtered scans.

The best way to illustrate this is with a test -- I just pushed a change which does the following:

1. Change this line to just `checkSizeLimit`
2. Adds a new test `testCheckLimitAfterFilteringRowCells`

If I comment out this checkSizeLimit, the added test fails -- the whole scan is able to complete in 1 rpc instead of the expected 4. So this illustrates that we need to have a size check here.

Personally I think it's also accurate to have a time limit check here, because for these types of scans I think they'd be able to circumvent our existing time limits. But within the scope of this JIRA, I can keep it to just size limit for now.